### DISCUSSION TITLE
Is there a way to get the value of a variable from a bash line of code?

### DISCUSSION THREAD

**Discussion by aborruso** [jupyter/notebook#6708] [Q&A]
Hi,
using this

```
a=!for i in {1..3};do\
  echo "$i";\
done

print(a)
``` 

I get the `a` list `['1', '2', '3']`.

Is there a way to get the `i` value and print it in python at the end of the loop?

Something like (that I know, it does not work):
 
```
!for i in {1..3};do\
  echo "$i";\
done

print(i)
``` 

A way to do it, is to split in two cells and use `%%bash --out`. Is there no way to write in a single cell long bash code and get a variable value to use it in some python code?

Thank you

![image](https://user-images.githubusercontent.com/30607/216545225-bfede188-d42e-432e-965f-45e4a0bf826f.png)

**Comment by drmrd** [comment]
The short answer is "no". IPython (and Jupyter Notebooks by extension) have no direct connection to the variables you're using in an external programming language called _via_ a magic (like `%%bash`).

When you assigned to `a` the value of `![for loop]` in your first example, it was populated with the list of lines printed to stdout during the execution of `[for loop]`. IPython is not tracking the value of the bash variable `i` during this process (nor to my knowledge would there be any way to do so reliably across systems).

Likewise, when you write `--out i` on a cell magic like in your screenshot, you are assigning the content echoed to stdout in your for loop to a new Python variable `i` (in this case as a multi-line string instead of a list). The fact that `i` was also the variable you used in your bash script is coincidental; IPython is not introspecting into the bash process to track the value of `i`. If you replace `--out i` with `--out anything` and the subsequent `print(i)` call with `print(anything)`, your notebook will function identically.

As to your last question about how to retrieve output from a bash script in one cell _via_ Jupyter/IPython wizardry, your `a = [for loop]` approach is the most straightforward that I've used. It should also be clear from the above statements that getting a list of stdout or stderr lines (rather than a specific variable value from the bash script) is as good as you're likely to get from a magic command. 

### Further Reading
- [The definition of the `--out` parameter in the `script` magic module.](https://github.com/ipython/ipython/blob/3339fad91de3e3da1f02b0b209afef20f021ca33/IPython/core/magics/script.py#L30-L43)
- [The `shebang` method of `ScriptMagics`](https://github.com/ipython/ipython/blob/3339fad91de3e3da1f02b0b209afef20f021ca33/IPython/core/magics/script.py#L166-L305) that actually executes `%%bash` (and similar) magic cells.
